File: win/system/backend/app/api/member_status.py
"""
Member status API endpoints
Handles custom status messages for members
"""

from typing import Dict, Any
from fastapi import APIRouter, HTTPException, Depends, Body

from app.services.status_service import (
    get_member_status, set_member_status, clear_member_status
)
from app.dependencies.auth import get_admin_user

router = APIRouter()

@router.get("/members/{member_identifier}/status")
async def get_member_status_endpoint(member_identifier: str):
    """Get status for a specific member (public endpoint)"""
    try:
        status = get_member_status(member_identifier)
        
        return {
            "success": True,
            "member_identifier": member_identifier,
            "status": status
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch member status: {str(e)}"
        )

@router.post("/members/{member_identifier}/status")
async def set_member_status_endpoint(
    member_identifier: str,
    status_data: Dict[str, Any] = Body(...),
    user=Depends(get_admin_user)
):
    """Set or update status for a member (admin only)"""
    try:
        status_text = status_data.get("text")
        emoji = status_data.get("emoji")
        
        if not status_text:
            raise HTTPException(
                status_code=400,
                detail="Status text is required"
            )
        
        # Validate status text length
        if len(status_text) > 100:
            raise HTTPException(
                status_code=400,
                detail="Status text must be 100 characters or less"
            )
        
        status = set_member_status(member_identifier, status_text, emoji)
        
        return {
            "success": True,
            "message": f"Status updated for {member_identifier}",
            "status": status
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to set member status: {str(e)}"
        )

@router.delete("/members/{member_identifier}/status")
async def clear_member_status_endpoint(
    member_identifier: str,
    user=Depends(get_admin_user)
):
    """Clear status for a member (admin only)"""
    try:
        success = clear_member_status(member_identifier)
        
        if success:
            return {
                "success": True,
                "message": f"Status cleared for {member_identifier}"
            }
        else:
            return {
                "success": False,
                "message": f"No status found for {member_identifier}"
            }
            
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to clear member status: {str(e)}"
        )

File: system/backend/src/routes/member_status.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Member status API endpoints.
 * Handles custom status messages for members.
 *
 * Relative paths — mounted under "/api" in main.ts, matching
 * app.include_router(member_status.router, prefix="/api") in the Python app.
 */

import { Router, type Request, type Response, type NextFunction } from 'express';

import { getMemberStatus, setMemberStatus, clearMemberStatus } from '../services/status_service.js';
import { requireAuth, requireAdmin } from '../dependencies/auth.js';
import { HttpError } from '../core/errors.js';

export const memberStatusRouter = Router();

/** Get status for a specific member (public endpoint) */
memberStatusRouter.get(
  '/members/:member_identifier/status',
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const memberIdentifier = req.params.member_identifier;
      const status = await getMemberStatus(memberIdentifier);

      res.json({ success: true, member_identifier: memberIdentifier, status });
    } catch (err) {
      next(new HttpError(500, `Failed to fetch member status: ${String(err)}`));
    }
  },
);

/** Set or update status for a member (admin only) */
memberStatusRouter.post(
  '/members/:member_identifier/status',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    const memberIdentifier = req.params.member_identifier;
    const statusData = req.body ?? {};
    const statusText = statusData.text;
    const emoji = statusData.emoji;

    if (!statusText) {
      next(new HttpError(400, 'Status text is required'));
      return;
    }

    // Validate status text length
    if (String(statusText).length > 100) {
      next(new HttpError(400, 'Status text must be 100 characters or less'));
      return;
    }

    try {
      const status = await setMemberStatus(memberIdentifier, statusText, emoji);

      res.json({
        success: true,
        message: `Status updated for ${memberIdentifier}`,
        status,
      });
    } catch (err) {
      next(new HttpError(500, `Failed to set member status: ${String(err)}`));
    }
  },
);

/** Clear status for a member (admin only) */
memberStatusRouter.delete(
  '/members/:member_identifier/status',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const memberIdentifier = req.params.member_identifier;
      const success = await clearMemberStatus(memberIdentifier);

      if (success) {
        res.json({ success: true, message: `Status cleared for ${memberIdentifier}` });
      } else {
        res.json({ success: false, message: `No status found for ${memberIdentifier}` });
      }
    } catch (err) {
      next(new HttpError(500, `Failed to clear member status: ${String(err)}`));
    }
  },
);